polkadot/node/network/approval-distribution/src/lib.rs

```diff
@@ -1961,6 +1961,15 @@ impl State {
 		}
 	}
 
+	// It is very important that aggression starts with oldest unfinalized block, rather than oldest
+	// unapproved block. We use approval checking lag just as a mean to trigger the aggression, but
+	// use finality lag to choose our aggression target.
+	//
+	// We have GRANDPA consensus around the last finalized block, but we don't for approvals.
+	// Using GRANDPA as source of truth, we can reason that an unfinalized but approved block will
+	// never be finalized because at least 1/3 of the nodes haven't seen enough approvals for it. It
+	// becomes mandatory that we share them in an aggressive manner wrt topology to increase the
+	// chance that the approval messages reach those nodes.
```

Comment:
I am not sure I fully understand this.

It is not really about chances, but more about making sure we are aggressing on the right thing (eventually).

Why doesn't this work with using approval voting lag and why does it work by using the oldest unfinalized block?

Let's consider using the approval voting lag: As with finality this is off-chain and we might not have full consensus on the current approval voting lag. (Some nodes have seen votes, others have not.)

Now let's assume that from our node's perspective with have an approval lag X, but others have an even larger lag X+n (because they had been restarted and have not seen votes, we have seen). Now we would do aggression on block X, resending votes the others don't need (yet), because they are still missing votes for X+1: Result, they will never receive the votes they actually need and we have a finality stall (if there are enough of those nodes).

Now why does it work by using the oldest non finalized block? Because it is a safe bet, only risking some additional work, but it can never happen that we are not sending something others might still need:

Observation: If something is finalized (somewhere), it must have been approved. This means if my node is lagging behind finality, I might start aggression on a block that actually has been finalized already, which is pointless, but harmless. It is already final and eventually my node will catch up, bumping up the finalized blocks and then continue aggression on the block that really was not yet finalized, sending the votes others actually need. -> Now we converge.

Reply:
Using the approval voting lag as a trigger should be safe, because even if we have divergent views and some nodes are not yet triggering aggression, although they should, this would only lead to the approval voting lag to grow more, until the trigger hits on enough nodes. 

So it is safe, the question is whether it is better than using the finality lag. I guess so, because we avoid triggering aggression levels in case finality lags for other reasons. Although what other reasons are there?

- Disputes: .... yes, but if something is disputed (and delaying finality) then it quite likely is also not approved (valid dispute).
- Disputes spam: Lot's of concluding valid disputes, might be slow to resolve and also delay finality, causing even more stress by triggering approval aggression levels. ... should be mitigated by disabling, but defense in depth is good in general.
- Network issues, which for some reason affect grandpa more than approvals.
- Other grandpa issue/malfunction

So there are slight benefits and the only downside I see is that some nodes might aggress a bit later than they would have otherwise, which might not even be a bad thing.

Reply:
> So it is safe, the question is whether it is better than using the finality lag. I guess so, because we avoid triggering aggression levels in case finality lags for other reasons. Although what other reasons are there?
> 
> * Disputes: .... yes, but if something is disputed (and delaying finality) then it quite likely is also not approved (valid dispute).
> * Disputes spam: Lot's of concluding valid disputes, might be slow to resolve and also delay finality, causing even more stress by triggering approval aggression levels. ... should be mitigated by disabling, but defense in depth is good in general.
> * Network issues, which for some reason affect grandpa more than approvals.
> * Other grandpa issue/malfunction
> 
> So there are slight benefits and the only downside I see is that some nodes might aggress a bit later than they would have otherwise, which might not even be a bad thing.

Yes, we don't want to trigger on finality lag, which could still happen due to other reasons (GRANDPA) while approval lag is optimal. I've seen this happen many times, which made the situation even worse since there would be a lot of additional network load for no good reason at all. 

I agree it is going to trigger a bit later compared to the triggering on finality lag, but the cost of that is much higher and I'd expect the same results of the aggression in both cases. Maybe the delay is also helpful as you said and prevents increasing the likely high load of the system.

Reply:
> Using the approval voting lag as a trigger should be safe, because even if we have divergent views and some nodes are not yet triggering aggression, although they should, this would only lead to the approval voting lag to grow more, until the trigger hits on enough nodes.

Not sure I follow this. 

Consider a scenario where half the nodes restart at the same time. And there was a candidate with all 30+ approvals from the half that didn't restart. The restarting half didn't also see the assignments for this candidate, so they don't see them as no-shows and don't trigger more tranches checks.
Now, the restarting half will have ever growing approval checking finality lag, but the other half won't. Why would it? In this scenario, using the finality lag as an aggression trigger is better as it always guarantees convergence at the cost of putting more stress when the network is already under stress.

Reply:
> I am not sure I fully understand this.
> It is not really about chances, but more about making sure we are aggressing on the right thing (eventually).
> 
> Why doesn't this work with using approval voting lag and why does it work by using the oldest unfinalized block?

Maybe I compressed it too much :). My comment assumes GRANDPA the source of consensus on what doesn't need more approval votes. Based on that, all other blocks in the chain(unfinalized) might be seen as unapproved by at least 1/3 of the nodes. So, without knowing which nodes know what about the first unfinalized block, we send out assignments/approvals aggressively to many peers hoping, that we get the block approved in more nodes so GRANDPA can advance finality.

> 
> Let's consider using the approval voting lag: As with finality this is off-chain and we might not have full consensus on the current approval voting lag. (Some nodes have seen votes, others have not.)
> 
> Now let's assume that from our node's perspective with have an approval lag X, but others have an even larger lag X+n (because they had been restarted and have not seen votes, we have seen). Now we would do aggression on block X, resending votes the others don't need (yet), because they are still missing votes for X+1: Result: They will never receive the votes they actually need and we have a finality stall (if there are enough of those nodes).
> 
> Now, why does it work by using the oldest non finalized block? Because it is a safe bet, only risking some additional work, but it can never happen that we are not sending something others might still need:
> 
> Observation: If something is finalized (somewhere), it must have been approved. This means if my node is lagging behind finality, I might start aggression on a block that actually has been finalized already, which is pointless, but harmless. It is already final and eventually my node will catch up, bumping up the finalized blocks and then continue aggression on the block that really was not yet finalized, sending the votes others actually need. -> Now we converge.

Yes this seems correct.

Reply:
> > Using the approval voting lag as a trigger should be safe, because even if we have divergent views and some nodes are not yet triggering aggression, although they should, this would only lead to the approval voting lag to grow more, until the trigger hits on enough nodes.
> 
> Not sure I follow this.
> 
> Consider a scenario where half the nodes restart at the same time. And there was a candidate with all 30+ approvals from the half that didn't restart. The restarting half didn't also see the assignments for this candidate, so they don't see them as no-shows and don't trigger more tranches checks. Now, the restarting half will have ever growing approval checking finality lag, but the other half won't. Why would it? In this scenario, using the finality lag as an aggression trigger is better as it always guarantees convergence at the cost of putting more stress when the network is already under stress.

The half which did not restart and have seen votes, would still see the block as unfinalized and the approach taken in this PR should be correct. They will start to aggress once the approval lag goes above threshold. They will aggress on that exact block for which approvals have never been seen by the restarting half.

Reply:
>  Now, the restarting half will have ever growing approval checking finality lag, but the other half won't. 

Yes, I think this is the most important statement. In this case, yes, those nodes wouldn't have any finality lag, so indeed, they would not trigger aggression.

Reply:
That should be fixed if we use req/response instead of gossip for aggression.

Reply:
> They will start to aggress once the approval lag goes above threshold. They will aggress on that exact block for which approvals have never been seen by the restarting half.

My point was they will not start aggressing (with approval checking lag as a trigger), even though it's them who should.

Reply:
:bulb: you are right @ordian . I jumped to conclusions too quickly. The nodes which did not restart, would never trigger aggression, as long as subsequent blocks still get approved by everybody, which should still happen. Indeed we should do something better than aggression, but for now, I think we need to switch the trigger back to finality lag.